src/scala/com/github/johnynek/bazel_deps/GeneratePom.scala

```diff
@@ -0,0 +1,107 @@
+        case Left(err) => Left(s"[ERROR]: Failed to parse ${path}.\n$err")
+      }
+    }
+  }
+
+  private def resolve[F[_]](model: Model, resolver: Resolver[F]): F[(Graph[MavenCoordinate, Unit])] = {
+    import resolver.resolverMonad
+
+    val deps = model.dependencies
+    resolver.buildGraph(deps.roots.toList.sorted, model)
+      .flatMap { graph =>
+        // This is a defensive check that can be removed as we add more tests
+        resolverMonad.catchNonFatal {
+          deps.roots.foreach { m => require(graph.nodes(m), s"$m") }
+          graph
+        }
+      }
+      .flatMap { graph =>
+        Normalizer(graph, deps.roots, model.getOptions.getVersionConflictPolicy) match {
+          case None =>
+            val output = graph.nodes.groupBy(_.unversioned)
+              .mapValues { _.map(_.version).toList.sorted }
+              .filter { case (_, s) => s.lengthCompare(1) > 0 }
+              .map { case (u, vs) => s"""${u.asString}: ${vs.mkString(", ")}\n""" }
+              .mkString("\n")
+            resolverMonad.raiseError(new Exception(
+              s"could not normalize versions:\n$output"))
+          case Some(normalized) =>
+
+            /**
+             * Make sure all the optional versioned items were found
+             */
+            val uvNodes = normalized.nodes.map(_.unversioned)
+            val check = deps.unversionedRoots.filterNot { u =>
+              uvNodes(u) || model.getReplacements.get(u).isDefined
+            }.toList match {
+              case Nil => resolverMonad.pure(())
+              case missing =>
+                val output = missing.map(_.asString).mkString(" ")
+                resolverMonad.raiseError(new Exception(s"Missing unversioned deps in the normalized graph: $output"))
+            }
+
+            for {
+              _ <- check
+            } yield (normalized)
+        }
+      }
+  }
+
+  def apply(deps: File, pomFile: File, groupId: String, artifactId: String, version: String): Unit = {
+    readModel(deps) match {
+      case Left(msg) => System.err.println(msg)
+      case Right(model) =>
+        val ec = scala.concurrent.ExecutionContext.Implicits.global
+        val resolver = new CoursierResolver(model.getOptions.getResolvers, ec, 3600.seconds)
+        val mavenCoordinateXml = resolver.run(resolve(model, resolver)) match {
+          case Failure(err) =>
+            Left(s"[ERROR]: Resolution failed ${err}.\n$err")
+          case Success((normalized)) =>
+            normalized.nodes.toList.map {
+              mv =>
+                mv.toXml
+            }
+        }
+
+        val pomXml = <project>
+          <modelVersion>4.0.0</modelVersion>
+
+          <groupId>{groupId}</groupId>
+          <artifactId>{artifactId}</artifactId>
+          <version>{version}</version>
+
+          <dependencies>
+            {mavenCoordinateXml}
+          </dependencies>
+        </project>
+
+        val p = new scala.xml.PrettyPrinter(80, 2)
```

Comment:
what about a method that returns the `p.format(pomXml)` so that we can write tests on this, then the apply method will call that then call the save method (which I am actually not clear how that is different from any saving of a string to a file).